CentaurusInfra/arktos
# SystemDaemonSet admission plugin allows daemonset created in system tenant only

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds the SysyemDaemonSet admission plugin which limits DaemonSet can only be created in system tenant by default (the plugin is enabled).

Current Arktos implementation has no constraints about DaemonSet creation; any tenant is able to have DaemonSet created. This lack of security is explicitly disallowed by the DaemonSet feature design in Arktos.

Though Arktos be enforces this check by default, it is still possible to disable this plugin so as to allow DaemonSet in any tenant, just as a standard kubrenetes does. This kind of behavior sometimes makes sense in Arktos Scale-up mode achieving higher k8s compatibility.  

**How to test**:
1. Start the local dev scale-up env by ```./hack/arktos-up.sh```
2. run following commands to create DaemonSet in system tenant, verify it is created
```
./cluster/kubectl.sh apply -f https://k8s.io/examples/controllers/daemonset.yaml 
./cluster/kubectl.sh get -n kube-system ds fluentd-elasticsearch
```
3. run following commands to create in user tenant, verify the request is rejected
```
./cluster/kubectl.sh create tenant user-test
./cluster/kubectl.sh apply --tenant user-test -f https://k8s.io/examples/controllers/daemonset.yaml 
Error from server (Forbidden): error when creating "daemonset.yaml": daemonsets.apps "fluentd-elasticsearch" is forbidden: only system tenant is allowed to have DaemonSet
```

**Does this PR introduce a user-facing change?**:

```release-note
YES - by default only system tenant is allowed to have DaemonSet
```
